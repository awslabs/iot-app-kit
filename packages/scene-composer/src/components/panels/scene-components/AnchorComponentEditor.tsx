import { AttributeEditor, FormField, Grid, Input, Select, SpaceBetween, TextContent } from '@awsui/components-react';
import { IconLookup, findIconDefinition } from '@fortawesome/fontawesome-svg-core';
import { debounce } from 'lodash';
import React, { useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { defineMessages, useIntl } from 'react-intl';

import { getGlobalSettings } from '../../../common/GlobalSettings';
import { SCENE_ICONS } from '../../../common/constants';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { COMPOSER_FEATURES, DefaultAnchorStatus, IValueDataBinding, SceneResourceType } from '../../../interfaces';
import { IAnchorComponentInternal, ISceneComponentInternal, useSceneDocument, useStore } from '../../../store';
import { shallowEqualsArray } from '../../../utils/objectUtils';
import { i18nSceneIconsKeysStrings } from '../../../utils/polarisUtils';
import { convertToIotTwinMakerNamespace, getSceneResourceInfo } from '../../../utils/sceneResourceUtils';
import { colors } from '../../../utils/styleUtils';
import { IComponentEditorProps } from '../ComponentEditor';
import { TextInput } from '../CommonPanelComponents';
import useDynamicScene from '../../../hooks/useDynamicScene';

import { ValueDataBindingBuilder } from './common/ValueDataBindingBuilder';
import { ColorSelectorCombo } from './tag-style/ColorSelectorCombo/ColorSelectorCombo';
import { DecodeSvgString } from './tag-style/ColorSelectorCombo/ColorSelectorComboUtils/DecodeSvgString';
import { IconPicker } from './tag-style/IconPicker/IconPicker';

export const convertParamsToKeyValuePairs = (params: Record<string, string>) => {
  return Object.keys(params).map((key) => {
    return { key, value: params[key] };
  });
};

const i18nIconStrings = defineMessages({
  [DefaultAnchorStatus.Info]: { defaultMessage: 'Info icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Warning]: { defaultMessage: 'Warning icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Error]: { defaultMessage: 'Error icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Video]: { defaultMessage: 'Video icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Custom]: { defaultMessage: 'Custom icon', description: 'Icon name label' },
});

export type IAnchorComponentEditorProps = IComponentEditorProps;

// TODO: This component is very poorly written, has too many responsibilities, and is using questionable hacks (like updating state on useEffect, and deouncing it to prevent a rerender cycle loop)
export const AnchorComponentEditor: React.FC<IAnchorComponentEditorProps> = ({
  node,
  component,
}: IAnchorComponentEditorProps) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const tagStyle = getGlobalSettings().featureConfig[COMPOSER_FEATURES.TagStyle];
  const updateComponentInternal = useStore(sceneComposerId)((state) => state.updateComponentInternal);
  const valueDataBindingProvider = useStore(sceneComposerId)(
    (state) => state.getEditorConfig().valueDataBindingProvider,
  );
  const anchorComponent = component as IAnchorComponentInternal;
  const [items, setItems] = useState<{ key: string; value: string; constraintText?: string }[]>([]);
  const hasDuplicateKeyRef = useRef<boolean>(false);
  const { listSceneRuleMapIds } = useSceneDocument(sceneComposerId);
  const intl = useIntl();

  const dynamicSceneEnabled = useDynamicScene();

  const ruleMapIds = listSceneRuleMapIds();
  const selectedRuleMapId =
    anchorComponent.ruleBasedMapId && ruleMapIds.includes(anchorComponent.ruleBasedMapId)
      ? anchorComponent.ruleBasedMapId
      : null;

  const onUpdateCallbackDebounced = useCallback(
    debounce(
      (componentPartial: any, replace?: boolean) => {
        const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
        updateComponentInternal(node.ref, componentPartialWithRef, replace);
      },
      dynamicSceneEnabled ? 1000 : 100,
    ), // TODO: Temporary solution for the error when updating entity too frequent. Will implement a better solution for GA.
    [node.ref, component.ref],
  );

  const onUpdateCallbackForChosenColor = useCallback(
    debounce((componentPartial: any, replace?: boolean) => {
      const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
      updateComponentInternal(node.ref, componentPartialWithRef, replace);
    }, 300),
    [node.ref, component.ref],
  );

  const onUpdateCallback = useCallback(
    (componentPartial: any, replace?: boolean) => {
      const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
      updateComponentInternal(node.ref, componentPartialWithRef, replace);
    },
    [node.ref, component.ref],
  );

  useMemo(() => {
    setItems(convertParamsToKeyValuePairs(anchorComponent.navLink?.params || {}));
  }, [anchorComponent]);

  useEffect(() => {
    // update the state if the items are changed
    const params = convertParamsToKeyValuePairs(anchorComponent.navLink?.params || {});
    if (!shallowEqualsArray(items, params)) {
      const updatedComponent = { ...anchorComponent } as IAnchorComponentInternal;
      updatedComponent.navLink = Object.assign({}, updatedComponent.navLink);
      if (!hasDuplicateKeyRef.current) {
        updatedComponent.navLink.params = Object.assign({}, Object.fromEntries(items.map((i) => [i.key, i.value])));
        onUpdateCallbackDebounced(updatedComponent, true);
      }
    }
  }, [items]);

  const updateItem = (itemIndex: number, newKey: string, newValue: string) => {
    const updatedItems = [...items];
    // checking duplicate key
    const keySet = new Set<string>();
    let keyConstraintText = '';
    hasDuplicateKeyRef.current = false;
    for (const item of items) {
      if (keySet.has(newKey)) {
        keyConstraintText = 'The key already exists.';
        hasDuplicateKeyRef.current = true;
        break;
      } else {
        keySet.add(item.key);
      }
    }

    updatedItems[itemIndex] = { key: newKey, value: newValue, constraintText: keyConstraintText };
    setItems(updatedItems);
  };

  const addItem = (key: string, value: string) => {
    const updatedItems = [...items, { key, value }];
    setItems(updatedItems);
  };

  const deleteItem = (itemIndex: number) => {
    const updatedItems = [...items];
    updatedItems.splice(itemIndex, 1);
    setItems(updatedItems);
  };

  const onUpdateNavLinkParams = useCallback(
    (index, prop: 'key' | 'value', value) => {
      const newKey = prop === 'key' ? value : items[index].key;
      const newValue = prop === 'value' ? value : items[index].value;
      updateItem(index, newKey, newValue);
    },
    [items],
  );

  const iconOptions = useMemo(() => {
    return Object.keys(SCENE_ICONS)
      .filter((icon) => (!tagStyle && icon !== 'Custom') || tagStyle)
      .map((sceneIcon) => {
        return {
          label: intl.formatMessage(i18nSceneIconsKeysStrings[sceneIcon]) || sceneIcon,
          value: sceneIcon,
        };
      });
  }, [tagStyle]);

  const iconSelectedOptionIndex = useMemo(() => {
    if (!anchorComponent.icon) {
      return -1;
    }

    const sceneResourceInfo = getSceneResourceInfo(anchorComponent.icon);
    return Object.keys(SCENE_ICONS).indexOf(sceneResourceInfo.value);
  }, [anchorComponent, iconOptions]);

  const iconString = useMemo(() => {
    const iconStrings = Object.values(SCENE_ICONS);
    if (iconSelectedOptionIndex < 0 || iconSelectedOptionIndex >= iconStrings.length) {
      return;
    }

    return btoa(iconStrings[iconSelectedOptionIndex]);
  }, [iconSelectedOptionIndex]);

  const ruleOptions = ruleMapIds
    .concat(
      selectedRuleMapId
        ? intl.formatMessage({
            defaultMessage: 'No Rule',
            description: 'signify No rule option to be selected in a drop down menu',
          })
        : [],
    )
    .map((ruleMapId) => ({ label: ruleMapId, value: ruleMapId }));

  const hasIcon = iconSelectedOptionIndex >= 0;
  const iconGridDefinition = hasIcon ? [{ colspan: 10 }, { colspan: 2 }] : [{ colspan: 12 }];
  const isCustomStyle = tagStyle && iconOptions[iconSelectedOptionIndex]?.value === 'Custom';
  const customIcon = (anchorComponent.customIcon ?? { prefix: 'fas', iconName: 'info' }) as IconLookup;

  return (
    <SpaceBetween size='s'>
      <FormField label={intl.formatMessage({ defaultMessage: 'Tag style', description: 'Form field label' })}>
        <Grid gridDefinition={iconGridDefinition}>
          <Select
            data-testid='anchor-default-icon-select'
            selectedOption={hasIcon ? iconOptions[iconSelectedOptionIndex] : null}
            onChange={(e) => {
              if (e.detail.selectedOption.value) {
                const icon = convertToIotTwinMakerNamespace(SceneResourceType.Icon, e.detail.selectedOption.value);
                onUpdateCallback({
                  icon,
                });
              }
            }}
            options={iconOptions}
            selectedAriaLabel={intl.formatMessage({
              defaultMessage: 'Selected',
              description:
                'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
            })}
            placeholder={intl.formatMessage({ defaultMessage: 'Choose an icon', description: 'placeholder' })}
          />
          {hasIcon &&
            (isCustomStyle ? (
              <DecodeSvgString
                selectedColor={anchorComponent.chosenColor ?? colors.infoBlue}
                iconString={iconString!}
                customIcon={findIconDefinition(customIcon)}
                width='32px'
                height='32px'
                ariaLabel={intl.formatMessage(i18nIconStrings[iconOptions[iconSelectedOptionIndex]?.value])}
              />
            ) : (
              <img
                aria-label={intl.formatMessage(i18nIconStrings[iconOptions[iconSelectedOptionIndex]?.value])}
                width='32px'
                height='32px'
                src={`data:image/svg+xml;base64,${iconString}`}
              />
            ))}
        </Grid>
      </FormField>
      {isCustomStyle && (
        <FormField>
          <SpaceBetween size='m'>
            <ColorSelectorCombo
              color={anchorComponent.chosenColor ?? colors.customBlue}
              onSelectColor={(pickedColor) => {
                onUpdateCallbackForChosenColor({
                  chosenColor: pickedColor,
                });
              }}
              onUpdateCustomColors={(chosenCustomColors) =>
                onUpdateCallbackForChosenColor({ customColors: chosenCustomColors })
              }
              customColors={anchorComponent.customColors}
              colorPickerLabel={intl.formatMessage({ defaultMessage: 'Color', description: 'Color' })}
              customColorLabel={intl.formatMessage({ defaultMessage: 'Custom colors', description: 'Custom colors' })}
            />
            <IconPicker
              onSelectIconChange={(pickedIcon) => {
                onUpdateCallback({
                  customIcon: pickedIcon,
                });
              }}
              selectedIcon={customIcon}
              iconPickerLabel={intl.formatMessage({ defaultMessage: 'Icon', description: 'Icon' })}
              iconFilterText={intl.formatMessage({ defaultMessage: 'Find icons', description: 'Find icons' })}
              iconFilterTextAriaLabel={intl.formatMessage({
                defaultMessage: 'Filter icons',
                description: 'Filter icons',
              })}
            />
          </SpaceBetween>
        </FormField>
      )}

      {valueDataBindingProvider && (
        <ValueDataBindingBuilder
          componentRef={anchorComponent.ref}
          binding={anchorComponent.valueDataBinding}
          valueDataBindingProvider={valueDataBindingProvider}
          onChange={(valueDataBinding: IValueDataBinding) => {
            // we don't want to merge the dataBindingContext, so we'll need to manually replace it
            const updatedComponent = { ...component, valueDataBinding };
            onUpdateCallback(updatedComponent, true);
          }}
        />
      )}

      <FormField label={intl.formatMessage({ defaultMessage: 'Rule Id', description: 'Form field label' })}>
        <Select
          data-testid='anchor-rule-id-select'
          selectedOption={selectedRuleMapId ? { label: selectedRuleMapId, value: selectedRuleMapId } : null}
          onChange={(e) => {
            const ruleMapId = e.detail.selectedOption.value;
            if (ruleMapId) {
              if (ruleMapIds.includes(ruleMapId)) {
                onUpdateCallbackDebounced({ ruleBasedMapId: ruleMapId });
              } else {
                onUpdateCallbackDebounced({ ruleBasedMapId: undefined });
              }
            }
          }}
          options={ruleOptions}
          selectedAriaLabel={intl.formatMessage({
            defaultMessage: 'Selected',
            description:
              'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
          })}
          disabled={ruleMapIds.length === 0}
          placeholder={intl.formatMessage({ defaultMessage: 'Choose a rule', description: 'placeholder' })}
        />
      </FormField>

      <FormField label={intl.formatMessage({ defaultMessage: 'Link Target', description: 'Form field label' })}>
        <TextInput
          data-testid='anchor-link-target-input'
          value={anchorComponent.navLink?.destination || ''}
          setValue={(e) => onUpdateCallback({ navLink: { destination: e } })}
        />
      </FormField>

      <TextContent>
        {intl.formatMessage({ defaultMessage: 'Link Parameters', description: 'Form section title' })}
      </TextContent>
      <AttributeEditor
        data-testid='anchor-attribute-editor-select'
        items={items}
        isItemRemovable={() => true}
        definition={[
          {
            label: intl.formatMessage({ defaultMessage: 'Key', description: 'Field label' }),
            control: (item, index) => (
              <Input
                value={item.key}
                placeholder={intl.formatMessage({ defaultMessage: 'Key', description: 'placeholder' })}
                onChange={({ detail }) => {
                  onUpdateNavLinkParams(index, 'key', detail.value);
                }}
              />
            ),
            constraintText: (item) => item.constraintText,
          },
          {
            label: intl.formatMessage({ defaultMessage: 'Value', description: 'Field label' }),
            control: (item, index) => (
              <Input
                value={item.value}
                placeholder={intl.formatMessage({ defaultMessage: 'Value', description: 'placeholder' })}
                onChange={({ detail }) => {
                  onUpdateNavLinkParams(index, 'value', detail.value);
                }}
              />
            ),
          },
        ]}
        addButtonText={intl.formatMessage({
          defaultMessage: 'Add parameter',
          description: 'Add parameter Button Text',
        })}
        removeButtonText={intl.formatMessage({
          defaultMessage: 'Remove parameter',
          description: 'Remove parameter Button Text',
        })}
        empty={intl.formatMessage({
          defaultMessage: 'No parameter',
          description: 'visual text to show there are no parameters',
        })}
        onAddButtonClick={() => addItem('', '')}
        onRemoveButtonClick={({ detail: { itemIndex } }) => {
          deleteItem(itemIndex);
        }}
      />
    </SpaceBetween>
  );
};
