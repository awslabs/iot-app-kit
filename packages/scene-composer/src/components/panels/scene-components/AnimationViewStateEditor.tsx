import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { FormField, Button, SpaceBetween, Autosuggest } from '@awsui/components-react';
import { useIntl } from 'react-intl';

import { IAnimationComponentInternal, useEditorState, ISceneNodeInternal } from '../../../store';
import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { animationObjectKey } from '../../three-fiber/AnimationComponent/AnimationComponent';

export type AnimationViewStateEditorProps = {
  node: ISceneNodeInternal;
  component: IAnimationComponentInternal;
  onUpdate: (currentAnimations: string[], noderef: string, component: IAnimationComponentInternal) => void;
};
export const AnimationViewStateEditor: React.FC<AnimationViewStateEditorProps> = ({
  node,
  component: animationComponent,
  onUpdate,
}: AnimationViewStateEditorProps) => {
  const { formatMessage } = useIntl();

  const sceneComposerId = useSceneComposerId();
  const { getObject3DBySceneNodeRef } = useEditorState(sceneComposerId);
  const object = getObject3DBySceneNodeRef(node.ref);
  const animationObject3D = object?.getObjectByName(animationObjectKey);
  const [selectedAnimations, setSelectedAnimations] = useState<string[] | undefined>();
  const animationPanelList = useRef<string[]>([]); // '' on end for next animation

  const animationOptions = useMemo(
    () =>
      animationObject3D?.animations.map((animation) => ({
        value: animation.name,
      })),
    [animationObject3D],
  );

  const updateAnimations = useCallback(
    (animation: string, i: number) => {
      if (selectedAnimations) {
        const currentAnimations = [...selectedAnimations];
        currentAnimations[i] = animation;
        setSelectedAnimations(currentAnimations);
        animationPanelList.current = [...currentAnimations, ''];
      }
    },
    [selectedAnimations],
  );

  const removeAnimation = useCallback(
    (i: number) => {
      if (selectedAnimations) {
        const currentAnimations = [...selectedAnimations];
        currentAnimations.splice(i, 1);
        setSelectedAnimations(currentAnimations);
        animationPanelList.current = [...currentAnimations, ''];
      }
    },
    [selectedAnimations],
  );

  useEffect(() => {
    const currentAnimations: string[] = [...animationComponent.currentAnimations] || [];
    setSelectedAnimations(currentAnimations);
    animationPanelList.current = [...currentAnimations, ''];
  }, [node.ref]);

  useEffect(() => {
    if (selectedAnimations && selectedAnimations != animationComponent.currentAnimations) {
      onUpdate(selectedAnimations, node.ref, animationComponent);
    }
  }, [selectedAnimations]);

  return (
    <SpaceBetween size='xs'>
      {animationPanelList.current.map((animation, i) => (
        <FormField
          key={`${i}-formField`}
          label={formatMessage({ defaultMessage: 'Choose Animations', description: 'FormField label' })}
        >
          {!!animation && (
            <Button
              key={`${i}-remove`}
              variant='link'
              data-testid={'removeButton' + i}
              onClick={() => removeAnimation(i)}
            >
              {formatMessage({
                defaultMessage: 'remove',
                description: 'label for the remove animation button',
              })}
            </Button>
          )}
          <Autosuggest
            key={`${i}-suggest`}
            onChange={({ detail }) => updateAnimations(detail.value, i)}
            value={animation}
            options={animationOptions}
            ariaLabel={formatMessage({
              defaultMessage: 'Autosuggest example with suggestions',
              description: 'Specifies the animation thats being selected',
            })}
            placeholder={formatMessage({
              defaultMessage: 'Enter Value',
              description: 'Default Message for before an animation is selected',
            })}
            empty={formatMessage({
              defaultMessage: 'No matches found',
              description: 'label for when the user searches for a nonexistent animation',
            })}
            enteredTextLabel={function (): string {
              return formatMessage({
                defaultMessage: 'No matches found',
                description: 'label for when the user searches for a nonexistent animation',
              });
            }}
          />
        </FormField>
      ))}
    </SpaceBetween>
  );
};
