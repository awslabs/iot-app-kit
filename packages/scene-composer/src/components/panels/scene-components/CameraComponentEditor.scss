.tm-camera-sub-form-label {
  padding: 10px 0px 0px 0px;
}
