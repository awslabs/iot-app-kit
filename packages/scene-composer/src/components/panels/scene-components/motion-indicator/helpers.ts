import { isEmpty } from 'lodash';

import { Component } from '../../../../models/SceneModels';
import { IMotionIndicatorComponentInternal } from '../../../../store';

export function updateComponentForColorTypeSelection(colorType: string, component: IMotionIndicatorComponentInternal) {
  const valueDataBindings = { ...component.valueDataBindings };
  const config = { ...component.config };
  if (colorType === Component.MotionIndicatorDataBindingName.BackgroundColor) {
    // move the foreground color data binding selections to be used by background color if available.
    if (!isEmpty(component.valueDataBindings[Component.MotionIndicatorDataBindingName.ForegroundColor])) {
      valueDataBindings[Component.MotionIndicatorDataBindingName.BackgroundColor] =
        valueDataBindings[Component.MotionIndicatorDataBindingName.ForegroundColor];
    } else {
      valueDataBindings[Component.MotionIndicatorDataBindingName.BackgroundColor] = {};
    }
    // clear the foreground color relatedd values since background color is preferred.
    valueDataBindings[Component.MotionIndicatorDataBindingName.ForegroundColor] = undefined;
    config.defaultBackgroundColor = config.defaultForegroundColor;
    config.defaultForegroundColor = undefined;
  } else {
    if (!isEmpty(component.valueDataBindings[Component.MotionIndicatorDataBindingName.BackgroundColor])) {
      valueDataBindings[Component.MotionIndicatorDataBindingName.ForegroundColor] =
        valueDataBindings[Component.MotionIndicatorDataBindingName.BackgroundColor];
    } else {
      valueDataBindings[Component.MotionIndicatorDataBindingName.ForegroundColor] = {};
    }
    valueDataBindings[Component.MotionIndicatorDataBindingName.BackgroundColor] = undefined;
    config.defaultForegroundColor = config.defaultBackgroundColor;
    config.defaultBackgroundColor = undefined;
  }
  const updatedComponent = { ...component, valueDataBindings, config };
  return updatedComponent;
}
