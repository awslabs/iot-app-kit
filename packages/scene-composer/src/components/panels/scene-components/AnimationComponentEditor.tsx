import React, { useCallback, useContext } from 'react';

import { accessStore, ISceneNodeInternal, IAnimationComponentInternal } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';

import { AnimationViewStateEditor } from './AnimationViewStateEditor';

export type AnimationEditorProps = {
  node: ISceneNodeInternal;
  component: IAnimationComponentInternal;
};

export const AnimationComponentEditor: React.FC<AnimationEditorProps> = ({ ...props }) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const updateComponentInternal = accessStore(sceneComposerId)((state) => state.updateComponentInternal);
  // istanbul ignore next
  const onUpdate = useCallback(
    (currentAnimations, noderef, component) => {
      updateComponentInternal(noderef, { ...component, currentAnimations });
    },
    [updateComponentInternal],
  );

  return <AnimationViewStateEditor {...props} onUpdate={onUpdate} />;
};
