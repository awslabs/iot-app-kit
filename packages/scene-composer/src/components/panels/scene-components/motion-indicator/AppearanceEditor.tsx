import React from 'react';
import { Box, FormField, Grid, Input, Select, SpaceBetween } from '@awsui/components-react';
import { isEmpty } from 'lodash';
import { useIntl } from 'react-intl';

import { IMotionIndicatorComponentInternal } from '../../../../store';
import { Component, Vector3 } from '../../../../models/SceneModels';

import { updateComponentForColorTypeSelection } from './helpers';
import ColorEditor from './ColorEditor';
import PreviewArrow from './PreviewArrow';

interface IAppearanceEditorProps {
  component: IMotionIndicatorComponentInternal;
  scale: Vector3;
  onUpdateCallback: (componentPartial: any, replace?: boolean | undefined) => void;
}

const AppearanceEditor: React.FC<IAppearanceEditorProps> = ({ component, onUpdateCallback }) => {
  const { formatMessage } = useIntl();
  const selectedColorType = component.valueDataBindings[Component.MotionIndicatorDataBindingName.ForegroundColor]
    ? Component.MotionIndicatorDataBindingName.ForegroundColor
    : Component.MotionIndicatorDataBindingName.BackgroundColor;
  const selectedForegroundColor = selectedColorType === Component.MotionIndicatorDataBindingName.ForegroundColor;

  const options = [
    {
      label: formatMessage({
        defaultMessage: 'Arrow with background',
        description: 'option text of a Select component',
      }),
      value: Component.MotionIndicatorDataBindingName.BackgroundColor,
    },
    {
      label: formatMessage({ defaultMessage: 'Arrow color', description: 'option text of a Select component' }),
      value: Component.MotionIndicatorDataBindingName.ForegroundColor,
    },
  ];

  return (
    <FormField label={formatMessage({ defaultMessage: 'Indicator appearance', description: 'Form Field label' })}>
      <SpaceBetween size='s'>
        <Grid gridDefinition={[{ colspan: 10 }, { colspan: 2 }]}>
          <Select
            data-testid='motion-indicator-color-type-select'
            selectedOption={options.find((elem) => elem.value === selectedColorType) || options[0]}
            onChange={(e) => {
              const value = e.detail.selectedOption.value;
              if (value && value !== selectedColorType) {
                const updatedComponent = updateComponentForColorTypeSelection(value, component);
                onUpdateCallback(updatedComponent, true);
              }
            }}
            options={options}
            selectedAriaLabel={formatMessage({
              defaultMessage: 'Selected',
              description:
                'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
            })}
            placeholder={formatMessage({ defaultMessage: 'Choose a color', description: 'placeholder' })}
          />
          {isEmpty(component.valueDataBindings[selectedColorType]) && (
            <PreviewArrow
              opacity={selectedForegroundColor ? 1 : component.config.backgroundColorOpacity}
              background={component.config.defaultBackgroundColor}
              color={selectedForegroundColor ? component.config.defaultForegroundColor : undefined}
            />
          )}
        </Grid>
        <FormField label={formatMessage({ defaultMessage: 'Arrow', description: 'Form Field label' })}>
          <Box display='inline'>
            {formatMessage({
              defaultMessage: '# of arrows',
              description: 'label for an input component selecting number of arrows',
            })}
            <div style={{ width: '37px', display: 'inline-block', marginLeft: '13px' }}>
              <Input
                data-testid='motion-indicator-arrow-number-input'
                inputMode='numeric'
                onChange={({ detail }) => {
                  const updatedComponent = {
                    ...component,
                    config: {
                      ...component.config,
                      numOfRepeatInY: Number(detail.value),
                    },
                  };
                  onUpdateCallback(updatedComponent, true);
                }}
                value={String(component.config.numOfRepeatInY)}
              />
            </div>
          </Box>
        </FormField>
        <ColorEditor component={component} selectedColorType={selectedColorType} onUpdateCallback={onUpdateCallback} />
      </SpaceBetween>
    </FormField>
  );
};

AppearanceEditor.displayName = 'AppearanceEditor';

export default AppearanceEditor;
