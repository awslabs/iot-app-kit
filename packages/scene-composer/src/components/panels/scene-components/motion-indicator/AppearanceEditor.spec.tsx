import React from 'react';
import { render } from '@testing-library/react';

import { mockComponent } from '../../../../../tests/components/panels/scene-components/MockComponents';
import { IMotionIndicatorComponentInternal } from '../../../../store';
import { KnownComponentType } from '../../../../interfaces';
import { Component } from '../../../../models/SceneModels';

import AppearanceEditor from './AppearanceEditor';
import { updateComponentForColorTypeSelection as mockUpdateComponentForColorTypeSelection } from './helpers';

jest.mock('./ColorEditor', () => (props: any) => <div id='ColorEditor' data-mocked='ColorEditor' {...props} />);

jest.mock('./PreviewArrow', () => (props: any) => (
  <div id='PreviewArrow' data-mocked='PreviewArrow'>
    {JSON.stringify(props)}
  </div>
));

jest.mock('./helpers', () => {
  return {
    ...jest.requireActual('./helpers'),
    updateComponentForColorTypeSelection: jest.fn(),
  };
});

describe('AppearanceEditor', () => {
  const component: IMotionIndicatorComponentInternal = {
    ...mockComponent,
    type: KnownComponentType.MotionIndicator,
    shape: Component.MotionIndicatorShape.LinearCylinder,
    valueDataBindings: {
      [Component.MotionIndicatorDataBindingName.BackgroundColor]: { ruleBasedMapId: 'rule-1' },
    },
    config: {
      numOfRepeatInY: 4,
      backgroundColorOpacity: 0.5,
    },
  };
  const onUpdateCallback = jest.fn();
  const mockUpdatedComponent = { ref: 'mock-1' };

  beforeEach(() => {
    jest.clearAllMocks();

    (mockUpdateComponentForColorTypeSelection as any).mockReturnValue(mockUpdatedComponent);
  });

  it('should render correctly without data binding', () => {
    const { container } = render(
      <AppearanceEditor component={component} scale={[1, 2, 3]} onUpdateCallback={onUpdateCallback} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with foreground color binding', () => {
    const { container } = render(
      <AppearanceEditor
        component={{
          ...component,
          valueDataBindings: {
            [Component.MotionIndicatorDataBindingName.ForegroundColor]: { ruleBasedMapId: 'rule-1' },
          },
        }}
        scale={[1, 2, 3]}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with background color binding', () => {
    const { container } = render(
      <AppearanceEditor
        component={{
          ...component,
          valueDataBindings: {
            [Component.MotionIndicatorDataBindingName.BackgroundColor]: { ruleBasedMapId: 'rule-1' },
          },
        }}
        scale={[1, 2, 3]}
        onUpdateCallback={onUpdateCallback}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
