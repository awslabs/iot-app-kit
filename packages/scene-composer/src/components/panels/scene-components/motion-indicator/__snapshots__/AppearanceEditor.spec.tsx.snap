// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppearanceEditor should render correctly with background color binding 1`] = `
<div>
  <div
    data-mocked="FormField"
    label="Indicator appearance"
  >
    <div
      data-mocked="SpaceBetween"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":10},{\\"colspan\\":2}]"
      >
        <div
          data-mocked="Select"
          data-testid="motion-indicator-color-type-select"
          options="[{\\"label\\":\\"Arrow with background\\",\\"value\\":\\"backgroundColor\\"},{\\"label\\":\\"Arrow color\\",\\"value\\":\\"foregroundColor\\"}]"
          placeholder="Choose a color"
          selectedarialabel="Selected"
          selectedoption="{\\"label\\":\\"Arrow with background\\",\\"value\\":\\"backgroundColor\\"}"
        />
      </div>
      <div
        data-mocked="FormField"
        label="Arrow"
      >
        <div
          data-mocked="Box"
          display="inline"
        >
          # of arrows
          <div
            style="width: 37px; display: inline-block; margin-left: 13px;"
          >
            <div
              data-mocked="Input"
              data-testid="motion-indicator-arrow-number-input"
              inputmode="numeric"
              value="4"
            />
          </div>
        </div>
      </div>
      <div
        component="[object Object]"
        data-mocked="ColorEditor"
        id="ColorEditor"
        selectedcolortype="backgroundColor"
      />
    </div>
  </div>
</div>
`;

exports[`AppearanceEditor should render correctly with foreground color binding 1`] = `
<div>
  <div
    data-mocked="FormField"
    label="Indicator appearance"
  >
    <div
      data-mocked="SpaceBetween"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":10},{\\"colspan\\":2}]"
      >
        <div
          data-mocked="Select"
          data-testid="motion-indicator-color-type-select"
          options="[{\\"label\\":\\"Arrow with background\\",\\"value\\":\\"backgroundColor\\"},{\\"label\\":\\"Arrow color\\",\\"value\\":\\"foregroundColor\\"}]"
          placeholder="Choose a color"
          selectedarialabel="Selected"
          selectedoption="{\\"label\\":\\"Arrow color\\",\\"value\\":\\"foregroundColor\\"}"
        />
      </div>
      <div
        data-mocked="FormField"
        label="Arrow"
      >
        <div
          data-mocked="Box"
          display="inline"
        >
          # of arrows
          <div
            style="width: 37px; display: inline-block; margin-left: 13px;"
          >
            <div
              data-mocked="Input"
              data-testid="motion-indicator-arrow-number-input"
              inputmode="numeric"
              value="4"
            />
          </div>
        </div>
      </div>
      <div
        component="[object Object]"
        data-mocked="ColorEditor"
        id="ColorEditor"
        selectedcolortype="foregroundColor"
      />
    </div>
  </div>
</div>
`;

exports[`AppearanceEditor should render correctly without data binding 1`] = `
<div>
  <div
    data-mocked="FormField"
    label="Indicator appearance"
  >
    <div
      data-mocked="SpaceBetween"
    >
      <div
        data-mocked="Grid"
        griddefinition="[{\\"colspan\\":10},{\\"colspan\\":2}]"
      >
        <div
          data-mocked="Select"
          data-testid="motion-indicator-color-type-select"
          options="[{\\"label\\":\\"Arrow with background\\",\\"value\\":\\"backgroundColor\\"},{\\"label\\":\\"Arrow color\\",\\"value\\":\\"foregroundColor\\"}]"
          placeholder="Choose a color"
          selectedarialabel="Selected"
          selectedoption="{\\"label\\":\\"Arrow with background\\",\\"value\\":\\"backgroundColor\\"}"
        />
      </div>
      <div
        data-mocked="FormField"
        label="Arrow"
      >
        <div
          data-mocked="Box"
          display="inline"
        >
          # of arrows
          <div
            style="width: 37px; display: inline-block; margin-left: 13px;"
          >
            <div
              data-mocked="Input"
              data-testid="motion-indicator-arrow-number-input"
              inputmode="numeric"
              value="4"
            />
          </div>
        </div>
      </div>
      <div
        component="[object Object]"
        data-mocked="ColorEditor"
        id="ColorEditor"
        selectedcolortype="backgroundColor"
      />
    </div>
  </div>
</div>
`;
